File: typescript/structural/adapter/hole_peg/adapter.ts
import {RoundPeg} from "./round_peg";
import {Peg} from "./peg";
import {SquarePeg} from './square_peg'
import {AngularPeg} from "./angular_peg";

export class Adapter extends Peg {
  private readonly peg: Peg

  constructor(peg: Peg) {
    super()
    this.peg = peg
  }

  getRadius(): number {
    if (this.peg instanceof RoundPeg) {
      return this.peg.getRadius()
    }
    if (this.peg instanceof SquarePeg) {
      return this.peg.getWidth() * Math.sqrt(2) / 2
    }
    if (this.peg instanceof AngularPeg) {
      const sides = this.peg.getSideLength()
      const a = sides[0]
      const b = sides[1]
      const c = sides[2]
      const p = (a + b + c) / 2

      return (a * b * c) / 4 * (Math.sqrt(p * (p - a) * (p - b) * (p - c)))
    }
    return 0
  }
}

File: dart/structural/adapter/hole_peg/bin/adapter.dart
import 'dart:math';

import './angular_peg.dart';
import './peg.dart';
import './round_peg.dart';
import './square_peg.dart';

class Adapter extends Peg {
  final Peg _peg;

  Adapter(this._peg);

  @override
  double getRadius() {
    if (_peg is RoundPeg) {
      return _peg.getRadius();
    }
    if (_peg is SquarePeg) {
      return (_peg as SquarePeg).getWidth() * sqrt(2) / 2;
    }
    if (_peg is AngularPeg) {
      final sides = (_peg as AngularPeg).getSideLength();
      final a = sides[0];
      final b = sides[1];
      final c = sides[2];
      final p = (a + b + c) / 2;

      return (a * b * c) / 4 * sqrt(p * (p - a) * (p - b) * (p - c));
    }
    return _peg.getRadius();
  }
}
